File: LeetCode_Algorithm/110/110.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
from typing import Optional
class Solution:
    # balanced tree
    # 모든 리프노드들이 같은 레벨을 가지도록 밸런스를 맞춰주는 트리
    # 좌우의 높이를 비교하여 1 이상 차이나면 false
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        root = self.makeTree(root, 0)
        return self.solve(root)[0]
        # bottom up 69ms 71%
    def solve(self, root):
        if root is None:
            return (True, -1)
        left_balanced, left_height = self.solve(root.left)
        right_balanced, right_height = self.solve(root.right)
        if left_balanced == False or right_balanced == False:
            return (False, 0)
        return ((abs(left_height-right_height) < 2), 1+max(left_height, right_height))

    def makeTree(self, root, idx):
        rootNode = TreeNode(root[idx])
        if 2*idx+1 < len(root): rootNode.left = self.makeTree(root, 2*idx+1)
        if 2*idx+2 < len(root): rootNode.right = self.makeTree(root,2*idx+2)
        return rootNode
print(Solution().isBalanced([3,9,20,None,None,15,7]))

File: LeetCode_Algorithm/110/110.js

function TreeNode(val, left, right) {
  this.val = (val===undefined ? 0 : val)
  this.left = (left===undefined ? null : left)
  this.right = (right===undefined ? null : right)
}

function makeTree(root, idx) {
  let rootNode = new TreeNode(root[idx])
  if (2*idx+1 < root.length) {rootNode.left = makeTree(root,2*idx+1)}
  if (2*idx+2 < root.length) {rootNode.right = makeTree(root,2*idx+2)}
  return rootNode 
}
function solve(root) {
  if (root == undefined) {return [true, -1]}
  let leftResult = solve(root.left); let rightResult = solve(root.right);
  if (leftResult[0] == false || rightResult[0] == false) {return [false, 0]}
  let r1 = Math.abs(leftResult[1]-rightResult[1]) < 2
  let r2 = 1+Math.max(leftResult[1], rightResult[1])
  return [r1, r2]
}
var isBalanced = function(root) {
  // 127ms 30%
  let rootNode = makeTree(root, 0)
  return solve(rootNode)[0]
};

console.log(isBalanced([3,9,20,null,null,15,7]))